tools/solidity_upgrade/UpgradeSuite.h

```diff
@@ -0,0 +1,96 @@
+/*
+	This file is part of solidity.
+
+	solidity is free software: you can redistribute it and/or modify
+	it under the terms of the GNU General Public License as published by
+	the Free Software Foundation, either version 3 of the License, or
+	(at your option) any later version.
+
+	solidity is distributed in the hope that it will be useful,
+	but WITHOUT ANY WARRANTY; without even the implied warranty of
+	MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+	GNU General Public License for more details.
+
+	You should have received a copy of the GNU General Public License
+	along with solidity.  If not, see <http://www.gnu.org/licenses/>.
+*/
+#pragma once
+
+#include "UpgradeChange.h"
+
+#include <liblangutil/ErrorReporter.h>
+
+#include <libsolidity/ast/ASTVisitor.h>
+#include <libsolidity/analysis/OverrideChecker.h>
+
+#include <regex>
+
+namespace dev
+{
+namespace solidity
+{
+namespace tools
+{
+
+/**
+ * The base upgrade module that can be inherited from. Doing so
+ * creates a basic upgrade module that facilitates access to
+ * change reporting.
+ */
+class Upgrade
+{
+public:
+	Upgrade(std::vector<UpgradeChange>& _changes): m_changes(_changes) {}
+
+protected:
+	/// A reference to a suite-specific set of changes.
+	/// It is passed to all upgrade modules and meant to collect
+	/// reported changes.
+	std::vector<UpgradeChange>& m_changes;
+};
+
+/**
+ * A specific upgrade module meant to be run after the analysis phase
+ * of the compiler.
+ */
+class AnalysisUpgrade: public Upgrade, public ASTConstVisitor
```

Comment:
@axic Regarding you question from the meeting: `Upgrade` can be easily inherited from in order to implement a `ParserUpgrade` or something like that. This could be used to find expressions that would need to be updated before the contract can be parsed successfully. This would either need a hook into the Solidity parser or some simple regex's.

Reply:
Once a mandatory upgrade is detected, it can use the current infrastructure to actually execute the change.